blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,141 @@
+---
+title: "Traffic Routing Demystified - Best practices for AKS Load Balancers"
+description: "Optimize your Traffic Routing in AKS with best practices, including externalTrafficPolicy. Learn how to preserve client source IPs, implement graceful pod termination, and ensure even traffic distribution across nodes for reliable production workloads."
+date: 2025-04-08
+authors:
+  - Mitch Shao
+  - Vaibhav Arora
+categories: general
+tags:
+  - traffic-management
+  - networking
+  - best practices
+  - load balancers
+---
+
+# Traffic Routing Demystified - Best practices for AKS Load Balancers
+
+Managing external traffic in Kubernetes clusters can be a complex task, especially when striving to maintain service reliability, optimize performance, and ensure seamless user experiences. With the increasing adoption of Kubernetes in production environments, understanding and implementing best practices for external traffic management when using the Azure Load Balancer has become essential.
+
+In this blog, we delve into the intricacies of Kubernetes’ `externalTrafficPolicy=Local` setting and explore strategies to gracefully handle pod shutdowns, rolling updates, and pod distribution. By following these best practices, you can enhance the resilience and reliability of your services while optimizing resource utilization across your AKS clusters.
+
+## `externalTrafficPolicy`: Cluster vs. Local
+
+**What does `externalTrafficPolicy` do?** In short, it influences how a cloud load balancer directs incoming traffic to Kubernetes nodes and how that traffic may hop between nodes to reach pods:
+
+- **Cluster Mode (default) (`externalTrafficPolicy=Cluster`)**:
+    All nodes in the cluster are included behind the Azure Load Balancer (ALB), regardless of whether they host any pods for the service. Incoming external traffic is distributed evenly across all nodes in the cluster. If a node receives traffic but doesn't have a local pod for the Service, Kubernetes (`kube-proxy`) will forward the request across the cluster network to a node that does. While this mode simplifies traffic distribution and reduces the risk of uneven load, it does not preserve the client’s original source IP.
```

Comment:
"it does not preserve the client’s original source IP" - sure, and the reason why most people would go to externalTrafficPolicy=Cluster is because they want the source IP, **but** the original thrust and key value of this work IMO is to point out that when this mode is used, the restart of any node for any reason will disrupt 1/N of your external traffic in a way which cannot but cause timeouts and connection resets for external clients.  So for example if you upgrade node images weekly for security reasons, you painfully disrupt all of your external connections once a week too.  That is a less well-known and also key reason to consider using externalTrafficPolicy=Local.

I'd consider trying to emphasize the above more in this post rather than doubling down on source IP.